src/apps/tenderize/common/tenderize.farm.contract-position-fetcher.ts

```diff
@@ -0,0 +1,51 @@
+import { Inject } from '@nestjs/common';
+
+import { APP_TOOLKIT, IAppToolkit } from '~app-toolkit/app-toolkit.interface';
+import { GetDataPropsParams, GetTokenBalancesParams } from '~position/template/contract-position.template.types';
+import {
+  SingleStakingFarmDataProps,
+  SingleStakingFarmTemplateContractPositionFetcher,
+} from '~position/template/single-staking.template.contract-position-fetcher';
+
+import { TenderizeContractFactory, TenderFarm } from '../contracts';
+
+import { TenderizeTokenDefinitionsResolver } from './tenderize.token-definition-resolver';
+
+export abstract class FarmTokenFetcher extends SingleStakingFarmTemplateContractPositionFetcher<
+  TenderFarm,
+  SingleStakingFarmDataProps
+> {
+  constructor(
+    @Inject(APP_TOOLKIT) protected readonly appToolkit: IAppToolkit,
+    @Inject(TenderizeTokenDefinitionsResolver)
+    private readonly tokenDefinitionsResolver: TenderizeTokenDefinitionsResolver,
+    @Inject(TenderizeContractFactory) protected readonly contractFactory: TenderizeContractFactory,
+  ) {
+    super(appToolkit);
+  }
+
+  getContract(address: string): TenderFarm {
+    return this.contractFactory.tenderFarm({ network: this.network, address });
+  }
+
+  async getFarmDefinitions() {
+    const definitions = await this.tokenDefinitionsResolver.getTokenDefinitions(this.network);
+    return definitions.map(v => ({
+      address: v.tenderFarm,
+      stakedTokenAddress: v.lpToken,
+      rewardTokenAddresses: [v.tenderToken],
+    }));
+  }
+
+  async getStakedTokenBalance({ address, contract }: GetTokenBalancesParams<TenderFarm, SingleStakingFarmDataProps>) {
+    return contract.stakeOf(address);
+  }
+
+  async getRewardTokenBalances({ address, contract }: GetTokenBalancesParams<TenderFarm, SingleStakingFarmDataProps>) {
+    return contract.availableRewards(address);
+  }
+
+  async getRewardRates(_: GetDataPropsParams<TenderFarm, SingleStakingFarmDataProps>) {
+    return [0];
```

Comment:
[Not blocking]
With a Reward Rate of 0, how can rewards are earned ? Do rewards get deposited in the contract on a specific period ?

Reply:
unfortunately, there is no function in their protocol to get the current reward rate, only the percentage of the stake that is redirected to liquidity-providing farmers when harvesting. apparently, to get the rewards rate, it would require a custom implementation per underlying staking protocol (livepeer, graph, matic, audius).

Reply:
i checked their subgraph and apparently, there's a way to calculate it. i will try and push a new commit with this feature.

Reply:
No worries, I was just curious. Good 👏  Job